File: Lume/src/compiler/lexer.ts
import { invalidOperation, mallformedFloat, operationType, syntaxError } from "../lib/error.ts";
import { util } from "../util.ts";

export enum tokenType {
    int,
    float,
    binOp,
    leftParen,
    rightParen,
};

export interface token {
    value : string;
    type : tokenType;
    pos : number;
}

export const binIdentifiers : string = "+-/*^%";
const letterIdentifiers : string = "qwertyuiopasdfghjklzxcvbnmQWERTYUIOPASDFGHJKLZXCVBNM";

export class lexer {
    private source : string;
    private carrotIndex : number;
    private tokens : token[];

    constructor(source : string) {
        this.source = source;
        this.carrotIndex = 0;
        this.tokens = [];
    }

    // moves the carrot ahead
    public shiftCarrot(steps : number) {
        this.carrotIndex += steps;
    }

    // reads the letter at a certain carrot index
    public readIndex(carrotPosition : number) : [token | undefined, number] {
        const char : string | undefined = this.source.at(carrotPosition);
        if (char === undefined) throw new Error("Undefined error at carrot position " + carrotPosition)
        if (char === " ") return [undefined, 1]

        if (util.isDigit(char) || char === "." && util.isDigit(this.source.at(carrotPosition + 1)) || char === "-" && util.isDigit(this.source.at(carrotPosition + 1))) {
            const numberExtends : string = this.getNumberExtends(carrotPosition);
            if (util.isFloat(numberExtends)) {
                return [{type : tokenType.float, value : numberExtends, pos : carrotPosition}, numberExtends.length];
            } else if(util.isInteger(numberExtends)) {
                return [{type : tokenType.int, value : numberExtends, pos : carrotPosition}, numberExtends.length];
            }
        } else if(binIdentifiers.includes(char)) {
            return [{type : tokenType.binOp, value : this.checkBinSyntax(carrotPosition), pos : carrotPosition}, 1];
        } else if(char === "(") {
            return [{type : tokenType.leftParen, value : char, pos : carrotPosition}, 1];
        } else if(char === ")") {
            return [{type : tokenType.rightParen, value : char, pos : carrotPosition}, 1];
        }

        return [undefined, 1];
    }

    // starts the tokenizing process for the code
    public tokenize() {
        while (this.carrotIndex < this.source.length && this.source.length > 0) {
            const tokenInformation : [token | undefined, number] = this.readIndex(this.carrotIndex);
            this.shiftCarrot(tokenInformation[1]);

            if (!tokenInformation[0]) continue;
            console.log(tokenInformation[0])
            this.tokens.push(tokenInformation[0])
        }
    }

    // returns the lexed tokens
    public getTokens() : token[] {
        return this.tokens;
    }

    // fetches the next carrots to get the full number
    private getNumberExtends(carrotPosition : number) : string {
        const slicedCharacter : string = this.source.slice(carrotPosition, this.source.length)
        let numberSnippits : string;
        // makes it work with negative numbers
        if (slicedCharacter.indexOf("-") == 0) {
            numberSnippits = slicedCharacter.slice(1).split(/[^\w\.]/)[0]; // split at each symbol apart from a dot
            numberSnippits = "-" + numberSnippits;
        } else {
            numberSnippits = slicedCharacter.split(/[^\w\.]/)[0];
        }

        const dotCount : null | string[] = numberSnippits.match(/\./g);

        if (numberSnippits[0] == ".") numberSnippits = "0" + numberSnippits;
        // numbers like 1.0.1
        if (dotCount !== null && dotCount.length >= 2) {
            new mallformedFloat(carrotPosition, carrotPosition + numberSnippits.length, "Float contains decimals " + numberSnippits, {
                genStackTrace : true,
                operationType : operationType.compiler
            })
        // numbers like 1a
        } else if(!util.isDigit(numberSnippits)) {
            if (util.isInteger(numberSnippits)) new mallformedFloat(carrotPosition, carrotPosition + numberSnippits.length, "Integer contains invalid character " + numberSnippits, {
                genStackTrace : true,
                operationType : operationType.compiler
            })
            if (util.isFloat(numberSnippits)) new mallformedFloat(carrotPosition, carrotPosition + numberSnippits.length, "Float contains invalid character " + numberSnippits, {
                genStackTrace : true,
                operationType : operationType.compiler
            })
        }

        const nextRightCharacter = this.source.slice(carrotPosition + numberSnippits.length, this.source.length).replaceAll(/[\s\n\r]/g, "") // replaces all space, \n and \r with space
        const nextleftCharacter = this.source.slice(0, carrotPosition).replaceAll(/[\s\n\r]/g, "")

        const nextToken : string | undefined = nextRightCharacter.at(1);
        const prevToken : string | undefined = nextleftCharacter.at(Math.max(nextleftCharacter.length - 1, 0));
        let prevPrevToken : string | undefined = nextleftCharacter.at(Math.max(nextleftCharacter.length - 2, 0));
        if (nextleftCharacter.length - 2 < 0) {
            prevPrevToken = undefined
        }

        // check, because -2 ^ 2 is not allowed
        if (nextToken === "^" && slicedCharacter.at(0) === "-") new invalidOperation(carrotPosition, carrotPosition + numberSnippits.length, "right side negative factor in unary operation", {
            genStackTrace : true,
            operationType : operationType.compiler
        })
        if (prevToken !== undefined && util.isDigit(prevToken) || prevToken !== undefined && !binIdentifiers.includes(prevToken) && (prevToken !== "(" || prevToken === "(" && prevPrevToken !== undefined && prevPrevToken !== "(" && !binIdentifiers.includes(prevPrevToken))) new syntaxError(carrotPosition, carrotPosition + numberSnippits.length, "Invalid unary operation. Expect binary operator, got " + prevToken, {
            genStackTrace : true,
            operationType : operationType.compiler
        })

        return numberSnippits;
    }

    // checks bin syntax for any errors
    // returns provided char for chaining
    private checkBinSyntax(carrotPosition : number) : string {
        const binOp : string | undefined = this.source.at(carrotPosition);
        if (!binOp) throw new Error("Carrot Position undefined at position " + carrotPosition);

        const nextRightCharacter = this.source.slice(carrotPosition, this.source.length).replaceAll(/[\s\n\r]/g, "") // replaces all space, \n and \r with space
        const nextleftCharacter = this.source.slice(0, carrotPosition).replaceAll(/[\s\n\r]/g, "")

        const nextToken : string | undefined = nextRightCharacter.at(1);
        const nextNextToken : string | undefined = nextRightCharacter.at(2);
        const prevToken : string | undefined = nextleftCharacter.at(Math.max(nextleftCharacter.length - 1, 0));

        if (nextToken !== undefined && nextToken === "(") return binOp;
        if (nextToken === undefined) new syntaxError(carrotPosition, carrotPosition + 1, "Right factor of binary operation empty", {
            genStackTrace : true,
            operationType : operationType.compiler
        })
        if (prevToken === undefined) new syntaxError(carrotPosition, carrotPosition + 1, "Left factor of binary operation empty", {
            genStackTrace : true,
            operationType : operationType.compiler
        })
        if (!util.isDigit(prevToken) && prevToken !== ")") new syntaxError(carrotPosition, carrotPosition + 1, "Left factor of binary operation invalid", {
            genStackTrace : true,
            operationType : operationType.compiler
        })
        if (!util.isDigit(nextToken) && nextToken === "-" && (nextNextToken === undefined || nextNextToken !== "(" && !util.isDigit(nextNextToken)) || !util.isDigit(nextToken) && nextToken !== "(" && nextToken !== "-") new syntaxError(carrotPosition, carrotPosition + 1, "Right factor of binary operation invalid", {
            genStackTrace : true,
            operationType : operationType.compiler
        })

        return binOp;
    }
}

File: Lume_Cpp/Src/Compiler/Lexer.cpp
#include "Lexer.hpp"

const char* identifiers[] = {
	// syntax crap
	"(",
	")",
	";",
	".",
	// bin operators
 	"+", "-", "*", 
	"/", "^", "%"
};

lexer::lexer(string fileContent) {
	this->sourceFile = fileContent;
	this->caretIndex = 0;
	this->tokens;
	this->fileLength = fileContent.length();
	this->currentToken = fileContent.at(0);
}

void lexer::advance(int steps) {
	this->caretIndex += steps;
	if (this->isAtEnd()) 
		return;
	this->currentToken = this->sourceFile.at(this->caretIndex);
}

const char lexer::getTokAtIndex(unsigned int index) {
	return this->sourceFile.at(index);
}

bool lexer::isAtEnd() {
	return this->caretIndex >= static_cast<int>(this->fileLength);
}

token lexer::createToken(tokenType type, string value) {
	token tok = token{
		value,
		type,
		this->caretIndex
	};
	return tok;
}

// return the lexed tokens if lexed correctly
vector<token> lexer::getLexedTokens() {
	return this->tokens;
}

// does the lexing and stuff
void lexer::lex() {
	if (this->fileLength == 0) return;

	while (!this->isAtEnd()) {
		lexer::scanResult scanRes = this->scanTokenAtIndex();
		this->advance(scanRes.strideSize);

		if (scanRes.tokenInst.getType() == tokenType::NULL_TOKEN) continue;
		this->tokens.push_back(scanRes.tokenInst);
	}
}

// scan the token at a current index
lexer::scanResult lexer::scanTokenAtIndex() {
	const char CT = this->currentToken;
	unsigned int CI = this->caretIndex;

	if (CT == NULL)
		throw runtime_error("Token Undefined at caret position " + this->caretIndex);
	if (CT == ' ')
		return lexer::scanResult{ createToken(tokenType::NULL_TOKEN, "NULL"), 1};

	if (util::isNumber(CT) 
		|| CT == '.' && util::isNumber(this->getTokAtIndex(CI + 1)) 
		|| CT == '-' && util::isNumber(this->getTokAtIndex(CI + 1))
		|| CT == '-' && this->getTokAtIndex(CI + 1) == '.' && util::isNumber(this->getTokAtIndex(CI + 2))) {
		string numExt(this->getNumExtends(CI));
		if (util::isFloat(&numExt)) {
			return lexer::scanResult{ createToken(tokenType::TT_float, numExt), static_cast<unsigned int>(numExt.length())};
		}
		else if (util::isInt(&numExt)) {
			return lexer::scanResult{ createToken(tokenType::TT_int16, numExt), static_cast<unsigned int>(numExt.length()) };
		}
	}
	else if (lexerHelper::isBinOp(CT)) {
		return lexer::scanResult{ createToken(tokenType::TT_binOp, lexer::checkBinSyntax(CI)), 1};
	}
	else if (CT == '(') {
		return lexer::scanResult{ createToken(tokenType::TT_leftParen, string(1,CT)), 1 };
	}
	else if (CT == ')') {
		return lexer::scanResult{ createToken(tokenType::TT_rightParen, string(1,CT)), 1 };
	}

	return lexer::scanResult{ createToken(tokenType::NULL_TOKEN, "NULL"), 1};
}

string lexer::getNumExtends(unsigned int index) {
	string croppedSource(this->sourceFile.substr(index));
	string numExt;

	// if number with minus, then cut it off becuase regex expr 
	// filters out syms, then add it back on later
	// otherwise just regex filter the text
	if (croppedSource.find_first_of("-") == 0) {
		string crop(croppedSource.substr(1));
		crop = lexerHelper::splitStrOnSym(&crop).at(0);
		crop.insert(0, "-");
		numExt = crop;
	}
	else {
		string cropBuffer(croppedSource);
		cropBuffer = lexerHelper::splitStrOnSym(&cropBuffer).at(0);
		numExt = cropBuffer;
	}

	__int64 tally = ranges::count(numExt, '.');

	if (numExt[0] == '.')
		numExt.insert(0, "0");
	else if (numExt[0] == '-'  && numExt[1] && numExt[1] == '.')
		numExt.insert(1, "0");
	
	if (tally > 1) {
		mallformedInteger(&index, index + static_cast<int>(numExt.length()), "Float contains multiple decimals '" + numExt + "'", errorConfig{
			true,
			processorType::LUME_compiler,
		});
	}
	else if (!util::isNumber(numExt.c_str())) {
		if (util::isInt(&numExt))
			mallformedFloat(&index, index + static_cast<int>(numExt.length()), "Float contains invalid character '" + numExt + "'", errorConfig{
				true,
				processorType::LUME_compiler,
				});
		if (util::isFloat(&numExt))
			mallformedInteger(&index, index + static_cast<int>(numExt.length()), "integer contains invalid character '" + numExt + "'", errorConfig{
				true,
				processorType::LUME_compiler,
			});
	}


	bool prevIsEmpty = false;
	string nextCharBuff(this->sourceFile.substr(index + numExt.length(), this->fileLength - (index + numExt.length())));
	string prevCharBuff;
	if (index == 0) {
		prevCharBuff = this->sourceFile;
		prevIsEmpty = true;
	}
	else {
		prevCharBuff = this->sourceFile.substr(0, index);
	}

	lexerHelper::condenseStr(&nextCharBuff);
	lexerHelper::condenseStr(&prevCharBuff);
	char nextTok = lexerHelper::atIndexIsValid(&nextCharBuff, 0)
		? nextCharBuff[0]
		: ' ';
	char prevTok = lexerHelper::atIndexIsValid(&prevCharBuff, max(static_cast<int>(prevCharBuff.length() - 1), 0))
		? prevCharBuff[max(static_cast<int>(prevCharBuff.length() - 1), 0)]
		: ' ';
	char prevSecTok = lexerHelper::atIndexIsValid(&prevCharBuff, max(static_cast<int>(prevCharBuff.length() - 2), 0))
		? prevCharBuff[max(static_cast<int>(prevCharBuff.length() - 2), 0)]
		: ' ';

	// check, because -2 ^ 2 is not allowed
	if (nextTok == '^' && croppedSource[0] == '-') {
		invalidOperation(&index, index + static_cast<int>(numExt.length()), "negative right side factor in unary operation", errorConfig{
			true,
			processorType::LUME_compiler,
			});
	}
	else if (prevCharBuff.length() - 2 < 0 || prevIsEmpty)
		return numExt;
	// checks if tokens allign correctly for an operation
	// It catches patterns like 1 1 or 1 ( 1
	// basically things that don't allign with 
	// standard operation terms
	else if (util::isNumber(prevTok)
		|| !lexerHelper::isBinOp(prevTok) && (prevTok != '('
		|| prevTok == '(' && prevSecTok != NULL && prevSecTok != '(' && !lexerHelper::isBinOp(prevSecTok))) {
		string str(1, prevTok);
		syntaxError(&index, index + static_cast<int>(numExt.length()), "Invalid operation. Expect binary operator, got '" + str + "'", errorConfig{
			true,
			processorType::LUME_compiler,
		});
	}

	return numExt;
}

string lexer::checkBinSyntax(unsigned int index) {
	if (!currentToken)
		throw runtime_error("Carrot Position undefined at position " + index);

	bool prevIsEmpty = false;
	string nextCharBuff(this->sourceFile.substr(index + 1, this->fileLength - (index + 1))); // plus once for the current char
	string prevCharBuff;

	if (index == 0) {
		prevCharBuff = this->sourceFile;
		prevIsEmpty = true;
	}
	else {
		prevCharBuff = this->sourceFile.substr(0, index);
	}

	lexerHelper::condenseStr(&nextCharBuff);
	lexerHelper::condenseStr(&prevCharBuff);
	int prevIndex = max(static_cast<int>(prevCharBuff.length() - 1), 0);
	char nextTok =	lexerHelper::atIndexIsValid(&nextCharBuff, 0) 
		? nextCharBuff[0]
		: ' ';
	char nextSecTok = lexerHelper::atIndexIsValid(&nextCharBuff, 1)
		? nextCharBuff[1]
		: ' ';
	char prevTok = lexerHelper::atIndexIsValid(&prevCharBuff, prevIndex)
		? prevCharBuff[prevIndex]
		: ' ';

	if (nextTok != ' ' && nextTok == '(')
		return string(1, currentToken);
	else if (nextTok == ' ')
		syntaxError(&index, index + 1, "Right factor of binary operation empty", errorConfig{
			true,
			processorType::LUME_compiler,
		});
	else if(prevTok == ' ') 
		syntaxError(&index, index + 1, "left factor of binary operation empty", errorConfig{
			true,
			processorType::LUME_compiler,
		});
	else if(!util::isNumber(prevTok) && prevTok != ')')
		syntaxError(&index, index + 1, "left factor of binary operation invalid", errorConfig{
			true,
			processorType::LUME_compiler,
		});
	else if(!util::isNumber(nextTok) && nextTok == '-' && (nextSecTok == ' '
		|| nextSecTok != '(' && !util::isNumber(nextSecTok))
		|| !util::isNumber(nextTok) && nextTok != '(' && nextTok != '-')
		syntaxError(&index, index + 1, "Right factor of binary operation invalid", errorConfig{
			true,
			processorType::LUME_compiler,
		});

	return string(1, currentToken);
}


////////////////////////////////////////////////////////
// Lexer helper funcs
////////////////////////////////////////////////////////


// checks if index provided fits within a string
bool lexerHelper::atIndexIsValid(string* source, unsigned int index) {
	return index >= static_cast<unsigned int>(source->length()) 
		? false 
		: true;
}

bool lexerHelper::isBinOp(char binOp) {
	string BID = BIN_IDENTIFIERS;
	return BID.find(binOp) != string::npos;
}

void lexerHelper::condenseStr(string* str) {
	str->erase(remove_if(str->begin(), str->end(), [](unsigned char _C) {
		return isspace(_C) || _C == '\n' || _C == '\r';
		}), str->end());
}

// splits string on every symbol and whitespace apart from dot
vector<string> lexerHelper::splitStrOnSym(string* str) {
	vector<string> res;
	string temp;
	for (const char _C : *str) {
		if (ispunct(_C) && _C != '.' || isspace(_C)) {
			res.push_back(temp);
			temp = "";
		}
		else {
			temp += string(1, _C);
		}
	}
	return res;
}